### ISSUE TITLE
support newInstance() with arguments

### ISSUE THREAD

**Issue by byteit101** [jruby/jruby#449]
Java cannot call newInstance on a ruby class with arguments (rubyClass.newInstance(object) will throw an exception)

**Comment by byteit101** [comment]
Context from JavaFX 7 jar dump:
```java
private static void launchApplication1(Class<? extends Application> paramClass, Class<? extends Preloader> paramClass1, String[] paramArrayOfString)
    throws Exception
  {
...
          Constructor localConstructor1 = paramClass1.getConstructor(new Class[0]);
          localPreloader = (Preloader)localConstructor1.newInstance(new Object[0]); /// THIS is where JavaFX fails to call ruby when paramClass is a JRuby class

          ParametersImpl.registerParameters(localPreloader, new ParametersImpl(paramArrayOfString));
```

If we pass javafx.Application::launch() a JRuby class (based on javafx.Application), it will fail at the above line in com.sun.javafx.application.LauncherImpl

**Comment by byteit101** [comment]
And here is the java 8 source, which is similar, but I've not tested it and it might work: http://hg.openjdk.java.net/openjfx/8/master/rt/file/f89b7dc932af/modules/graphics/src/main/java/com/sun/javafx/application/LauncherImpl.java#l791

**Comment by byteit101** [comment]
Here is a quick test case and the full error source:
Original test case:
```rb
require 'jruby/core_ext'
class App2 < javafx.application.Application; def start(stage); end; end
App2.become_java!
javafx.application.Application.launch(App2.java_class.to_java(java.lang.Class))
```
That case error:
```
Exception in Application constructor
Java::JavaLang::RuntimeException: Unable to construct Application instance: class javafx.application.Application
	from com.sun.javafx.application.LauncherImpl.launchApplication1(com/sun/javafx/application/LauncherImpl.java:907)
	from com.sun.javafx.application.LauncherImpl.lambda$launchApplication$155(com/sun/javafx/application/LauncherImpl.java:182)
	from java.lang.Thread.run(java/lang/Thread.java:745)

```

Reduced case:
```rb
require 'jruby/core_ext'
class App2 < javafx.application.Application; def start(stage); end; end
App2.become_java!
App2.java_class.to_java(java.lang.Class).getConstructor([].to_java(java.lang.Class)).newInstance([].to_java java.lang.Object)
```

The error I get
```
Java::JavaLang::InstantiationException: 
	from java.lang.reflect.Constructor.newInstance(java/lang/reflect/Constructor.java:423)
	from java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:498)
	from RUBY.<eval>((irb):2)
	from org.jruby.RubyKernel.eval(org/jruby/RubyKernel.java:976)
	from org.jruby.RubyKernel.loop(org/jruby/RubyKernel.java:1287)
	from org.jruby.RubyKernel.catch(org/jruby/RubyKernel.java:1096)
	from org.jruby.RubyKernel.catch(org/jruby/RubyKernel.java:1096)
	from home.patrick.$_dot_rvm.rubies.jruby_minus_9_dot_0_dot_5_dot_0.bin.irb.<top>(/home/patrick/.rvm/rubies/jruby-9.0.5.0/bin/irb:13)
	from java.lang.invoke.MethodHandle.invokeWithArguments(java/lang/invoke/MethodHandle.java:627)
```

JavaFX Source with error handling:
```java
try {
                        Constructor<? extends Application> c = appClass.getConstructor();
                        app.set(c.newInstance());
                        // Set startup parameters
                        ParametersImpl.registerParameters(app.get(), new ParametersImpl(args));
                        PlatformImpl.setApplicationName(appClass);
                    } catch (Throwable t) {
                        System.err.println("Exception in Application constructor");
                        constructorError = t;
                        error = true;
                    }

...
if (constructorError != null) {
                    String msg = "Unable to construct Application instance: " + appClass;
                    if (!notifyError(msg, constructorError)) {
                        throw new RuntimeException(msg, constructorError);
                    }
                }
```

**Comment by headius** [comment]
I'm still not clear what needs to be done here and this is a bit invasive so I think we're going to bump it.

@byteit101 I need to sit down with you and @kares and figure out exactly what we need to do here. I know the class extension stuff is SUPER complicated...so it will be a challenge.

**Comment by headius** [comment]
Removing from any release. @byteit101 We'd still like to work with you on this at some point.

**Comment by byteit101** [comment]
Sure, I've just been busy recently

**Comment by kares** [comment]
looked into an issue and its now getting a bit ... well complicated :) Java sub-classes (byte-code) generate a proxy class while `become_java!` does kind of the same for 'pure' Ruby classes. the first hit is that `become_java!` does not allow Java proxy class hierarchies - it's probably unintentional but kind of wanted since by removing the restriction the generated class is not 100% functional. 
might try some more experiments to get the two play together.

**Comment by byteit101** [comment]
I was looking at this again recently after #6141 , along with what I think is a duplicate #5270 and related #2369

I have a bit more time these days, with the pandemic and all. As such, @headius I'd like to take up the offers above:
> @byteit101 I need to sit down with you and @kares and figure out exactly what we need to do here. I know the class extension stuff is SUPER complicated...so it will be a challenge.

> @byteit101 We'd still like to work with you on this at some point.

I would like to, at least for the jrubyfx perspective, support simple, no-arg ctors. I have a very hacky solution running locally, but it's very hacky (oh, and did I mention how hacky it is?). Supporting arg ctors is.. way more complicated...

I made a diagram of how I think object initialization of concrete extended classes works at a relatively high level with the new java side tacked on. (blue = java, red = ruby, the orange document = the joint java/ruby object). The lazy init is exclusively to solve #2369. The ??? for the initialize has super() could be several things, from api changes (some examples below), to fibers/continuations, to really magical syntactic sugar breaking `#initialize` methods in to two. No idea how to do the latter, or even if it's possible. I _think_ that we only need a pre-super init method that returns the super arguments, and a post-super init method. Is that accurate?

![jruby object init](https://user-images.githubusercontent.com/577344/94491146-a68e8b00-01b5-11eb-8d25-0a6d288d08c9.png)

A bunch of (terrible) API ideas (vs fiber or syntactic sugar options):
```rb
class A < com.MyAbstract
  java_signature "A(int a, String b, bool c)"
  def java_init(a, b c)
    return [a+5, b.reverse, !c] # calls super with these args.
    # No obvious way to pass computations from java_init -> initialize outside of ivars
  end
  def initialize(a, b, c) # original a, b, c
    # super is prohibited here, throws some error or is a no-op
  end
end

class A < com.MyAbstract
  # Not just a nicer def, but also allows overrides to be more obvious
  java_compute_super "(int a, String b, bool c)" do |a, b c|
    return [a+5, b.reverse, !c] # calls super with these args
  end
end

class A < com.MyAbstract
# add to the become_java call
  become_java! do |(a, b, c), parent|
    parent.call(a + 5, b.reverse, !c)
  end
end

class A < com.MyAbstract
  java_init_passthrough true # always calls super(*args) at the start
end

# not pictured:  muck with initialize call, and enforce first/last/something?
```end

**Comment by byteit101** [comment]
Oh, and I didn't look into normal reified classes, but the jrubyc compiled classes seem to have a very different pattern to how they execute

**Comment by byteit101** [comment]
Fixed by #6422
